File: m_exporter/poc/mini-prometheus/go/snapshot.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"sort"
	"time"
)

const snapshotVersion = 1

type snapshotFile struct {
	Version int               `json:"version"`
	SavedAt int64             `json:"saved_at_ms"`
	Series  []snapshotSeries  `json:"series"`
}

type snapshotSeries struct {
	Labels  []Label      `json:"labels"`
	Samples [][2]float64 `json:"samples"` // [t_ms, value]
}

func (st *Storage) Snapshot() ([]byte, error) {
	st.mu.RLock()
	out := snapshotFile{
		Version: snapshotVersion,
		SavedAt: time.Now().UnixMilli(),
		Series:  make([]snapshotSeries, 0, len(st.series)),
	}
	for _, s := range st.series {
		s.mu.RLock()
		samples := make([][2]float64, 0, s.size)
		for i := 0; i < s.size; i++ {
			idx := (s.head - s.size + i + s.capacity) % s.capacity
			samples = append(samples, [2]float64{float64(s.samples[idx].T), s.samples[idx].V})
		}
		out.Series = append(out.Series, snapshotSeries{
			Labels:  []Label(s.Labels),
			Samples: samples,
		})
		s.mu.RUnlock()
	}
	st.mu.RUnlock()
	return json.Marshal(out)
}

func (st *Storage) Restore(b []byte) (int, error) {
	var snap snapshotFile
	if err := json.Unmarshal(b, &snap); err != nil {
		return 0, err
	}
	if snap.Version != snapshotVersion {
		return 0, fmt.Errorf("unsupported snapshot version: %d", snap.Version)
	}
	for _, ss := range snap.Series {
		lbls := Labels(ss.Labels)
		sort.Sort(lbls)
		s := st.getOrCreate(lbls)
		for _, sa := range ss.Samples {
			s.Append(int64(sa[0]), sa[1])
		}
	}
	return len(snap.Series), nil
}

// WriteSnapshot serializes storage to path atomically (tmp + rename).
func WriteSnapshot(path string, st *Storage) error {
	data, err := st.Snapshot()
	if err != nil {
		return err
	}
	tmp := path + ".tmp"
	if err := os.WriteFile(tmp, data, 0o644); err != nil {
		return err
	}
	return os.Rename(tmp, path)
}

File: m_exporter/poc/mini-prometheus/rust/src/snapshot.rs
use std::path::Path;

use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};

use crate::storage::Storage;

const SNAPSHOT_VERSION: u32 = 1;

#[derive(Serialize, Deserialize)]
struct SnapshotFile {
    version: u32,
    saved_at_ms: i64,
    series: Vec<SnapshotSeries>,
}

#[derive(Serialize, Deserialize)]
struct SnapshotSeries {
    labels: Vec<SnapshotLabel>,
    samples: Vec<(i64, f64)>,
}

#[derive(Serialize, Deserialize)]
struct SnapshotLabel {
    name: String,
    value: String,
}

pub fn write(path: &Path, storage: &Storage) -> Result<()> {
    let snap = SnapshotFile {
        version: SNAPSHOT_VERSION,
        saved_at_ms: chrono::Utc::now().timestamp_millis(),
        series: storage.iter_series_for_snapshot(|labels, samples| SnapshotSeries {
            labels: labels
                .0
                .iter()
                .map(|l| SnapshotLabel {
                    name: l.name.clone(),
                    value: l.value.clone(),
                })
                .collect(),
            samples: samples.iter().map(|s| (s.t, s.v)).collect(),
        }),
    };
    let data = serde_json::to_vec(&snap).context("encode snapshot")?;
    let tmp = path.with_extension("json.tmp");
    std::fs::write(&tmp, &data).with_context(|| format!("write {}", tmp.display()))?;
    std::fs::rename(&tmp, path)
        .with_context(|| format!("rename {} -> {}", tmp.display(), path.display()))?;
    Ok(())
}

pub fn read(path: &Path, storage: &Storage) -> Result<usize> {
    let data = std::fs::read(path).with_context(|| format!("read {}", path.display()))?;
    let snap: SnapshotFile = serde_json::from_slice(&data).context("decode snapshot")?;
    if snap.version != SNAPSHOT_VERSION {
        anyhow::bail!("unsupported snapshot version: {}", snap.version);
    }
    let n = snap.series.len();
    for ss in snap.series {
        let mut metric = String::new();
        let mut extra: Vec<(String, String)> = Vec::new();
        for l in ss.labels {
            if l.name == "__name__" {
                metric = l.value;
            } else {
                extra.push((l.name, l.value));
            }
        }
        if metric.is_empty() {
            continue;
        }
        for (t, v) in ss.samples {
            storage.append(&metric, &extra, t, v);
        }
    }
    Ok(n)
}
